text/0000-bitfields.md

```diff
@@ -0,0 +1,340 @@
+
+# Summary
+[summary]: #summary
+
+Add support for bitfields via a single per-field attribute.
+
+# Motivation
+[motivation]: #motivation
+
+Bitfields are commonly encountered when interacting with C code. Since many of
+the details of the behavior of bitfields are left to the implementation, it is
+hard to write cross-platform rust code that uses bitfields correctly.
+
+Currently, such code often manually collapses bitfields into a single integer
+field and attempts to extract and set the individual fields via bitshifts. This
+approach works for simple bitfields and when the full set of target platforms is
+known, however, once the bitfields become more complicated or if the code is
+compiled for an unexpected target, such code will silently fail. In particular,
+code that assumes that bitfields are placed into the integer in increasing bit
+order will likely break on big-endian machines.
+
+Appendix A contains various examples of the (to some) surprising behavior of
+bitfields on the x86-64 linux platform.
+
+This rfc introduces a single attribute that can be applied to fields inside of
+`#[repr(C)]` structs to turn them into bitfields.
+
+# Detailed design
+[design]: #detailed-design
+
+The attribute `#[bitfield(N)]` is added. Here, `N` can be any constant
+expression that evaluates to a value of type `usize`.
+
+The attribute can be applied to fields inside a `#[repr(C)]` struct which have
+any of following types:
+
+* `u8`, `u16`, `u32`, `u64`, `usize`,
+* `i8`, `i16`, `i32`, `i64`, `isize`.
+
+For example:
+
+```rust
+#[repr(C)]
+pub struct perf_event_attr {
+    /* ... */
+
+    #[bitfield(1)] pub disabled: u64,
+    #[bitfield(1)] pub inherit: u64,
+    #[bitfield(1)] pub pinned: u64,
+    #[bitfield(1)] pub exclusive: u64,
+    #[bitfield(1)] pub exclude_user: u64,
+    #[bitfield(1)] pub exclude_kernel: u64,
+    #[bitfield(1)] pub exclude_hv: u64,
+    #[bitfield(1)] pub exclude_idle: u64,
+    #[bitfield(1)] pub mmap: u64,
+    #[bitfield(1)] pub comm: u64,
+    #[bitfield(1)] pub freq: u64,
+    #[bitfield(1)] pub inherit_stat: u64,
+    #[bitfield(1)] pub enable_on_exec: u64,
+    #[bitfield(1)] pub task: u64,
+    #[bitfield(1)] pub watermark: u64,
+    #[bitfield(1)] pub precise_ip: u64,
+    #[bitfield(1)] pub mmap_data: u64,
+    #[bitfield(1)] pub sample_id_all: u64,
+    #[bitfield(1)] pub exclude_host: u64,
+    #[bitfield(1)] pub exclude_guest: u64,
+    #[bitfield(1)] pub exclude_callchain_kernel: u64,
+    #[bitfield(1)] pub exclude_callchain_user: u64,
+    #[bitfield(1)] pub mmap2: u64,
+    #[bitfield(1)] pub comm_exec: u64,
+    #[bitfield(39)] pub __reserved_1: u64,
+
+    /* ... */
+}
+```
+
+`N` must evaluate to an integer in the range `[0, sizeof(T)]` where `T` is the
+declared type of the field. It specifies the width of the field.
+
+Many details of the behavior of such fields are left to the implementation.
+However, the implementation is encouraged to follow the behavior of the dominant
```

Comment:
Not just encouraged, but probably should be required, since if there's no guarantee it becomes useless for FFI which will likely be the primary consumer of bitfields.